edge/pkg/devicetwin/dmiserver/server.go

```diff
@@ -0,0 +1,242 @@
+	"k8s.io/klog/v2"
+
+	beehiveContext "github.com/kubeedge/beehive/pkg/core/context"
+	beehiveModel "github.com/kubeedge/beehive/pkg/core/model"
+	deviceconst "github.com/kubeedge/kubeedge/cloud/pkg/devicecontroller/constants"
+	"github.com/kubeedge/kubeedge/cloud/pkg/devicecontroller/types"
+	"github.com/kubeedge/kubeedge/common/constants"
+	messagepkg "github.com/kubeedge/kubeedge/edge/pkg/common/message"
+	"github.com/kubeedge/kubeedge/edge/pkg/common/modules"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dmiclient"
+	"github.com/kubeedge/kubeedge/edge/pkg/devicetwin/dtcommon"
+	"github.com/kubeedge/kubeedge/edge/pkg/metamanager/dao"
+	"github.com/kubeedge/kubeedge/pkg/apis/devices/v1alpha2"
+	pb "github.com/kubeedge/kubeedge/pkg/apis/dmi/v1"
+)
+
+const (
+	SockPath = "/etc/kubeedge/dmi.sock"
+	Limit    = 1000
+	Burst    = 100
+)
+
+type server struct {
+	limiter  *rate.Limiter
+	dmiCache *DMICache
+}
+
+type DMICache struct {
+	MapperMu        *sync.Mutex
+	DeviceMu        *sync.Mutex
+	DeviceModelMu   *sync.Mutex
+	MapperList      map[string]*pb.MapperInfo
+	DeviceModelList map[string]*v1alpha2.DeviceModel
+	DeviceList      map[string]*v1alpha2.Device
+}
+
+func (s *server) MapperRegister(ctx context.Context, in *pb.MapperRegisterRequest) (*pb.MapperRegisterResponse, error) {
+	if !s.limiter.Allow() {
+		return nil, fmt.Errorf("fail to register mapper because of too many request: %s", in.Mapper.Name)
+	}
+
+	if in.Mapper.Protocol == "" {
+		klog.Errorf("fail to register mapper %s because the protocol is nil", in.Mapper.Name)
+		return nil, fmt.Errorf("fail to register mapper %s because the protocol is nil", in.Mapper.Name)
+	}
+
+	klog.V(4).Infof("receive mapper register: %+v", in.Mapper)
+	err := saveMapper(in.Mapper)
+	if err != nil {
+		klog.Errorf("fail to save mapper %s to db with err: %v", in.Mapper.Name, err)
+		return nil, err
+	}
+	s.dmiCache.MapperMu.Lock()
+	s.dmiCache.MapperList[in.Mapper.Name] = in.Mapper
+	s.dmiCache.MapperMu.Unlock()
+
+	if !in.WithData {
+		return &pb.MapperRegisterResponse{}, nil
+	}
+
+	var deviceList []*pb.Device
+	var deviceModelList []*pb.DeviceModel
+	for _, device := range s.dmiCache.DeviceList {
+		protocol, err := dtcommon.GetProtocolNameOfDevice(device)
+		if err != nil {
+			klog.Errorf("fail to get protocol name with err: %+v", err)
+			continue
+		}
+
+		if protocol == in.Mapper.Protocol {
+			dev, err := dtcommon.ConvertDevice(device)
+			if err != nil {
+				klog.Errorf("fail to convert device %s with err: %v", device.Name, err)
+				continue
+			}
+
+			name, ok := s.dmiCache.DeviceModelList[device.Spec.DeviceModelRef.Name]
+			if !ok {
+				klog.Errorf("fail to get device model %s in deviceModelList", s.dmiCache.DeviceModelList[device.Spec.DeviceModelRef.Name])
+				continue
+			}
+			dm, err := dtcommon.ConvertDeviceModel(name)
+			if err != nil {
+				klog.Errorf("fail to convert device model %s with err: %v", s.dmiCache.DeviceModelList[device.Spec.DeviceModelRef.Name], err)
```

Comment:
`klog.Errorf("fail to convert device model %s with err: %v", device.Spec.DeviceModelRef.Name, err)`
%s is not proper

Reply:
done. make it `device.Spec.DeviceModelRef.Name` and `%s` is proper.